src/v/kafka/server/handlers/handler_probe.cc

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "kafka/server/handlers/handler_probe.h"
+
+#include "config/configuration.h"
+#include "kafka/server/handlers/handler_interface.h"
+#include "prometheus/prometheus_sanitize.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/metrics.hh>
+
+#include <chrono>
+#include <memory>
+
+namespace kafka {
+
+handler_probe_manager::handler_probe_manager()
+  : _probes(max_api_key() + 2)
+  , _unknown_handler_index(max_api_key() + 1) {
+    for (size_t i = 0; i < _probes.size(); i++) {
+        auto key = api_key{i};
+
+        if (handler_for_key(key) || i == _unknown_handler_index) {
+            _probes[i] = std::make_unique<handler_probe>(key);
+            _probes.at(i)->setup_metrics();
+        }
+    }
+}
+
+handler_probe& handler_probe_manager::get_probe(api_key key) {
+    if (!handler_for_key(key)) {
+        key = api_key{_unknown_handler_index};
+    }
+
+    return *_probes.at(key);
+}
+
+handler_probe::handler_probe(api_key key)
+  : _key(key) {}
+
+void handler_probe::setup_metrics() {
+    namespace sm = ss::metrics;
+
+    if (config::shard_local_cfg().disable_metrics()) {
+        return;
+    }
+
+    const char* handler_name;
+    if (auto handler = handler_for_key(_key)) {
+        handler_name = handler.value()->name();
+    } else {
+        handler_name = "unknown_handler";
+    }
+
+    std::vector<sm::label_instance> labels{sm::label("handler")(handler_name)};
+    auto aggregate_labels = config::shard_local_cfg().aggregate_metrics()
+                              ? std::vector<sm::label>{sm::shard_label}
+                              : std::vector<sm::label>{};
+
+    _metrics.add_group(
+      prometheus_sanitize::metrics_name("kafka:handler"),
+      {
+        sm::make_counter(
+          "requests_completed",
+          [this] { return _requests_completed; },
+          sm::description("Number of requests completed"),
```

Comment:
Should say "kafka" and "of the given API type" or soemthing like that. Same for other metrics.

Keep in mind that the autogen dashboard uses for the chart title so things like "requests" can have many meanings across redpanda and we have confusing cases like "size of the database" which actually only applies to the KV store: because the group name "kafka:handler" in this case does not appear in the dashboard title.

Reply:
Switching the names to `kafka_*_per_handler` for all the metrics.

Reply:
Is the metric group name not in the generated title? If that's the case it makes my suggestions not ideal, but hopefully they still help.

"handler" is a label, I don't think it needs to be in the name, too.

Reply:
The metric group name is not in the title or otherwise exposed in the UI which is the crux of the problem. From our point of view it seems redundant to add "kafka" when the group name already says Kafka but for reasonable auto-generated dashboards it is necessary.

Reply:
@ballard26 just to clarify here I was talking about the description (line 75), not the name. The description must stand alone w/o reference to the metric group or name due to the way we use it in autogenerated dashboards.